src/components/swarm/fa.rs

```diff
@@ -0,0 +1,128 @@
+            gamma,
+            id: PhantomId::default(),
+        }
+    }
+
+    pub fn new_with_id<P>(
+        alpha: f64,
+        beta: f64,
+        gamma: f64,
+    ) -> Box<dyn Component<P>>
+        where
+            P: LimitedVectorProblem<Element = f64>,
+    {
+        Box::new(Self::from_params(alpha, beta, gamma))
+    }
+}
+
+impl FireflyPositionsUpdate<Global> {
+    pub fn new<P>(alpha: f64, beta: f64, gamma: f64) -> Box<dyn Component<P>>
+        where
+            P: LimitedVectorProblem<Element = f64>,
+    {
+        Self::new_with_id(alpha, beta, gamma)
+    }
+}
+
+impl<P, I> Component<P> for FireflyPositionsUpdate<I>
+    where
+        P: LimitedVectorProblem<Element = f64>,
+        I: Identifier,
+{
+    fn init(&self, _problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        state.insert(RandomizationParameter(self.alpha));
+        Ok(())
+    }
+
+    fn execute(&self, problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        let mut populations = state.populations_mut();
+        let mut rng = state.random_mut();
+
+        // Prepare parameters
+        let &Self {
+            beta, gamma, ..
+        } = self;
+        let a = state.get_value::<RandomizationParameter>();
+
+        // Get necessary state
+        let mut individuals = populations.current_mut();
+        let mut evaluator = state.borrow::<Evaluator<P>>();
+
+        // scale for adapting to problem domain
+        let scales = problem.domain()
+            .iter()
+            .map(|p| (p.end - p.start).abs())
+            .collect::<Vec<f64>>();
+
+
+        // Perform the update step.
+        // compare all individuals
+        for i in  0..individuals.len() {
+            for j in 0..individuals.len() {
+                // if individual j is "more attractive" (i.e. has lower fitness), move towards j
+                if individuals[i].objective() > individuals[j].objective() {
+                    // draw random values from uniform distribution between 0 and 1
+                    // according to paper: also possible to use normal distribution, depending on problem
+                    let rands: Vec<f64> = (0..problem.dimension()).map(|_| rng.gen_range(0.0..1.0)).collect();
+                    let mut current = individuals[i].clone();
+                    izip!(current.solution_mut(), individuals[j].solution(), &scales, rands)
+                        .map(|(xi, xj, scale, rand)| {
+                            // calculate "attractiveness"
+                            //let b = ;
+                            // calculate value that should be added to current position
+                            let pos = beta * (-gamma * (*xi - xj).powf(2.0)).exp() * (xj - *xi) + a * (rand - 0.5) * scale;
+                            (xi, pos) })
+                        .for_each(|(xi, pos)| *xi += pos);
+                    individuals[i] = current;
+
+                    evaluator.evaluate(problem, state, from_mut(&mut individuals[i]));
```

Comment:
I think the problem here is that the `state` is still borrowed because of the `individuals` when calling `evaluator.evaluate`. Seems like we have to pop the population and push it again at the end to avoid the borrow. This kind of thing is still awkward in mahf.

Reply:
I tried, but the compiler still tells me:
![grafik](https://github.com/mahf-opt/mahf/assets/18464215/a72f8195-03b0-4bec-a40e-2256e524630d)

Reply:
Oh right. This is probably not very intuitive, but the `Evaluator` only wraps a `Box<dyn Evaluate>`, which we can access with `as_inner_mut`. And the `Evaluate` trait requires `&mut self`, so we have to borrow the `Evaluator` mutably. In fact, we need to remove it entirely (either with `State::take` and `State::insert` at the end or wrap the nested loop in `State::holding::<Evaluator>`), because the `state` would otherwise be borrowed by the `Evaluator` itself, which we need to pass it to.

You can also take a look at the implementation of the `PopulationEvaluator`, which does all this correctly.

Reply:
> You can also take a look at the implementation of the PopulationEvaluator, which does all of this correctly.

That's what I did but it didn't work. I pushed my recent approach. I will probably try `State::take` and `State::insert` next.

Reply:
You forgot to call `as_inner_mut`. Also, unfortunately, you must replace each call to `rng` with `state.random_mut()` to avoid borrowing the `state`.